src/main/cpp/native_glue.cc

```diff
@@ -19,140 +19,72 @@
+//   art/runtime/check_jni.cc:65]     in call to NewStringUTF
+// See:
+//   https://stackoverflow.com/questions/35519823/jni-detected-error-in-application-input-is-not-valid-modified-utf-8-illegal-st
+// The strategy here is to use standard C functions to convert the UTF-8 directly to UTF-16, which Java handles nicely.
+// The following two functions are taken from this repo:
+//   https://github.com/incanus/android-jni/blob/master/app/src/main/jni/JNI.cpp#L57-L86
+
+jstring litecore::jni::UTF8ToJstring(JNIEnv *env, char *s, size_t size) {
+    std::u16string ustr = std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>().from_bytes(s, s+size);
+
+    jstring jstr = env->NewString(reinterpret_cast<const jchar *>(ustr.c_str()), ustr.size());
+    if (jstr == nullptr) {
+        C4Error error = {LiteCoreDomain, kC4ErrorMemoryError, 0};
+        throwError(env, error);
     }
 
-    *output = newBytes;
-    return newStrLen;
-}
-
-void litecore::jni::ModifiedUTF8CharToUTF8(char *input) {
-    char c = input[0];
-    char c1 = input[1] & 0x3F;
-    char c2 = input[2] & 0x3F;
-    char d = input[3];
-    char d1 = input[4] & 0x3F;
-    char d2 = input[5] & 0x3F;
-
-    int surrogate[2] = { ((c & 0x0F) << 12) | (c1 << 6) | c2, ((d & 0x0F) << 12) | (d1 << 6) | d2 };
-    int codePoint = ((surrogate[0] - 0xD800) << 10) + (surrogate[1] - 0xDC00) + 0x10000;
-
-    input[0] = 0xF0 | (codePoint  >> 18);
-    input[1] = 0x80 | ((codePoint >> 12) & 0x3F);
-    input[2] = 0x80 | ((codePoint >> 6) & 0x3F);
-    input[3] = 0x80 | ((codePoint & 0x3F));
+    return jstr;
 }
 
-void litecore::jni::ModifiedUTF8ToUTF8(char *input) {
-    size_t len = 0;
-    size_t i = 0;
-    bool needsModify = false;
-    for(i = 0; input[i] != 0; i++) {
-        if(input[i] == '\xed') {
-            // According to https://docs.oracle.com/javase/1.5.0/docs/guide/jni/spec/types.html#wp16542
-            // modified UTF-8 for codepoints > FFFF will always start with 0xED
-            needsModify = true;
-            ModifiedUTF8CharToUTF8(&input[i]);
-            i += 5;
-        }
+std::string litecore::jni::JstringToUTF8(JNIEnv *env, jstring jstr) {
+    jsize len = env->GetStringLength(jstr);
+    if (len < 0) {
+        C4Error error = {LiteCoreDomain, kC4ErrorInvalidParameter, 0};
+        throwError(env, error);
+        return std::string();
     }
 
-    if(!needsModify) {
-        return;
+    const jchar *chars = env->GetStringChars(jstr, nullptr);
+    if (chars == nullptr) {
+        C4Error error = {LiteCoreDomain, kC4ErrorMemoryError, 0};
+        throwError(env, error);
+        return std::string();
     }
 
-    len = i;
-    size_t j = 0;
-    for(i = 0; input[i] != 0; i++) {
-        if((uint8_t)input[i] >= 0xF0) {
-            i+= 3;
-            size_t j;
-            for(j = i + 1; j < len - 2; j++) {
-                input[j] = input[j+2];
-            }
+    auto str = std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>()
```

Comment:
All these `wstring_convert` calls can throw C++ exceptions, at least `bad_alloc`. Is there a `catch` block somewhere in the call chain?

Reply:
Fixed.